File: l-umi/Week_42/BOJ_1051_숫자_정사각형.kt
import java.io.*
import java.util.*

fun main() {
   val br = BufferedReader(InputStreamReader(System.`in`))
   var st = StringTokenizer(br.readLine())
   
   val N = st.nextToken().toInt()
   val M = st.nextToken().toInt()
   
   ///////////////////////////////////////
   
   val board = Array(N) { IntArray(M) }
   for(i in 0 until N) {
       val line = br.readLine()
       for(j in 0 until M) {
           // 문자를 숫자로 변환 ('1' -> 1로 변환하기 위해 '0'을 빼줌)
           board[i][j] = line[j] - '0'
       }
   }
   
   // 특정 크기의 정사각형이 가능한지 확인하는 함수
   fun checkSize(size: Int): Boolean {
       for(i in 0..N-size) {
           for(j in 0..M-size) {
               // 네 꼭짓점이 모두 같은지 확인
               if(board[i][j] == board[i][j+size-1] && 
                  board[i][j] == board[i+size-1][j] && 
                  board[i][j] == board[i+size-1][j+size-1]) {
                   return true
               }
           }
       }
       return false
   }
   
   ///////////////////////////////////////   
   val maxSize = minOf(N, M)
   
   for(size in maxSize downTo 1) {
       if(checkSize(size)) {
           println(size * size)
           break
       }
   }
}

File: l-umi/Week_42/BOJ_1051_숫자_정사각형.c
#include <stdio.h>

#define MAX_SIZE 50

// size의 정사각형이 가능한지 확인하는 함수
int checkSize(int board[MAX_SIZE][MAX_SIZE], int N, int M, int size) {
    for(int i = 0; i <= N-size; i++) {
        for(int j = 0; j <= M-size; j++) {
            if(board[i][j] == board[i][j+size-1] && 
               board[i][j] == board[i+size-1][j] && 
               board[i][j] == board[i+size-1][j+size-1]) {
                return 1;
            }
        }
    }
    return 0;
}

int main() {
    int N, M;
    scanf("%d %d", &N, &M);

    int board[MAX_SIZE][MAX_SIZE];
    char line[MAX_SIZE + 1]; // 문자열 입력을 위한 배열
    
    for(int i = 0; i < N; i++) {
        scanf("%s", line);
        for(int j = 0; j < M; j++) {
            board[i][j] = line[j] - '0';
        }
    }
    
    /////////////////////////////////////////
    
    int maxSize = N < M ? N : M;
    
    for(int size = maxSize; size >= 1; size--) {
        if(checkSize(board, N, M, size)) {
            printf("%d", size * size);
            break;
        }
    }
    
    return 0;
}